src/validation.cpp

```diff
@@ -3284,6 +3288,121 @@ void CChainState::ProcessLoanEvents(const CBlockIndex* pindex, CCustomCSView& ca
+    const auto premium{COIN + rewardPct};
+
+    std::map<DCT_ID, CFuturesPrice> futuresPrices;
+    CDataStructureV0 tokenKey{AttributeTypes::Token, 0, TokenKeys::DFIP2203Disabled};
+
+    cache.ForEachLoanToken([&](const DCT_ID& id, const CLoanView::CLoanSetLoanTokenImpl& loanToken) {
+        tokenKey.typeId = id.v;
+        const auto disabled = attributes->GetValue(tokenKey, false);
+        if (disabled) {
+            return true;
+        }
+
+        const auto useNextPrice{false}, requireLivePrice{true};
+        const auto discountPrice = cache.GetAmountInCurrency(discount, loanToken.fixedIntervalPriceId, useNextPrice, requireLivePrice);
+        const auto premiumPrice = cache.GetAmountInCurrency(premium, loanToken.fixedIntervalPriceId, useNextPrice, requireLivePrice);
+        if (!discountPrice || !premiumPrice) {
+            return true;
+        }
+
+        futuresPrices.emplace(id, CFuturesPrice{*discountPrice, *premiumPrice});
+
+        return true;
+    });
+
+    const uint32_t startHeight = pindex->nHeight - blockPeriod;
+    std::map<CFuturesUserKey, CFuturesUserValue> unpaidContracts;
+    cache.ForEachFuturesUserValues([&](const CFuturesUserKey& key, const CFuturesUserValue& futuresValues){
+        if (key.height <= startHeight) {
+            return false;
+        }
+
+        const auto source = cache.GetLoanTokenByID(futuresValues.source.nTokenId);
+        assert(source);
+
+        if (source->symbol == "DUSD") {
+            const DCT_ID destId{futuresValues.destination};
+            const auto destToken = cache.GetLoanTokenByID(destId);
+            assert(destToken);
+
+            try {
+                const auto& premiumPrice = futuresPrices.at(destId).premium;
+                if (premiumPrice > 0) {
+                    const auto total = DivideAmounts(futuresValues.source.nValue, premiumPrice);
+                    cache.AddBalance(key.owner, {destId, total});
+                }
+            } catch (const std::out_of_range&) {
+                unpaidContracts.emplace(key, futuresValues);
+            }
+
+        } else {
+            const auto tokenDUSD = cache.GetToken("DUSD");
+            assert(tokenDUSD);
+
+            try {
+                const auto& discountPrice = futuresPrices.at(futuresValues.source.nTokenId).discount;
+                const auto total = MultiplyAmounts(futuresValues.source.nValue, discountPrice);
+                cache.AddBalance(key.owner, {tokenDUSD->first, total});
+            } catch (const std::out_of_range&) {
+                unpaidContracts.emplace(key, futuresValues);
+            }
+        }
+
+        return true;
+    }, {static_cast<uint32_t>(pindex->nHeight), {}, std::numeric_limits<uint32_t>::max()});
+
+    const auto resVal = GetFutureSwapContractAddress();
+    assert(resVal);
+
+    CDataStructureV0 liveKey{AttributeTypes::Live, ParamIDs::Economy, EconomyKeys::DFIP2203Tokens};
+    auto balances = attributes->GetValue(liveKey, CBalances{});
```

Comment:
currently we are only tracking the amounts that are send into the contract right? no tracking of paid out funds.
Might risk another AtomicSwap-like-leak?

Reply:
History of paid out funds added. Will add an RPC call in a new PR.

Reply:
awesome. thank you. (I cant find how to resolve discussions. I think I don't have the rights....)